### DISCUSSION TITLE
Type inner component

### DISCUSSION THREAD

**Discussion by tavoyne** [emotion-js/emotion#2686] [Q&A] [answered]
```tsx
export interface TestProps {
  Icon: (props: { className: string }) => JSX.Element;
  children?: string;
}

export default function Test({ Icon, children }: TestProps) {
  return (
    <div>
      <Icon css={{ width: "2rem" }} />
      {children}
    </div>
  );
}
```

Gives me the following error (on `Icon`):
```
Property 'className' is missing in type '{ css: { width: string; }; }' but required in type '{ className: string; }'. ts(2741)
```

Anybody knows what's the correct way to type this?

**Comment by Andarist** [comment]
Please always try to share a repro case in a runnable form - either by providing a git repository to clone or a codesandbox. OSS maintainers usually can't afford the time to set up the repro, even if exact steps are given.

**Reply by tavoyne** [reply]
It's not a bug. I'm just asking the community for help, not trying to bother you.

**Reply by Andarist** [reply]
Ye, no worries. But even if it's not a bug it's usually easier for maintainers to get to your question if runnable repro is provided. That is because it's easy to play around with it, tweak it and try things out if needed.

**Comment by srmagura** [comment, marked as answer]
You just need to set `className` to optional in the definition of the `Icon` prop. So `className?: string`.

**Reply by tavoyne** [reply]
Thanks. Yes indeed. 

But since the `css` prop is always set, the `className` will never be `undefined`, so it shouldn't be necessary, don't you think?

**Reply by Andarist** [reply]
There might be room for improvement in this type:
https://github.com/emotion-js/emotion/blob/2d3d7dd1d8b0de9dbfd50e42ee716ac00fb70ecd/packages/react/types/jsx-namespace.d.ts#L5-L9

I'm slightly worried that this could lead to some perf problems but I'm not sure - the whole thing is somewhat magical. If I understand correctly this would require using `Omit` on the input `P`, like [here](https://www.typescriptlang.org/play?#code/PQKhAIAECsGcA9x0eEwBQAzArgOwMYAuAlgPa5IIAUAlOAN4C+66hAngA4Cm4AItgFsBbAMIBlMQDUAhgBtwAXgbh8sWAC5wAIlIAjaFvCNwLdt3AB1YoQAWI8gBNrZXHPFiACgCdSHADweAHyK2viy0moActICXIZc8IRcuA6w4ADWXGykmOAe6ODgAPzgsIRexLgA5uAJSSlpHgDaWmERsNGxhgBkGVk5eQC6BYXFeS1tUTFxg7WJyaml5ZVVI6NjAPIC1gEANOAA5JMd0wfBvfQqapr8QqISMvLM64WaHuAXV7BFN4LC7o8jGtwG8Rm8ANwsBxcNpeHhVWR6OQMEauWKwDjSfA8ABSYgAGij1pUkl5MFieABRWRcWK4QgMZ6jElcMkU8DU2nJQgicJqRnAlls7Ecml0wgAQUIy102CSsG8vlZ7AFxPprPJIs54pENmIsgccNwUplcp4TGBZh4ABliLovNIvGwALLSVxVLgOE0VWXyvwifZBEJWWz2FLOchuCSK-xBSFq0mangASXpFVwsGI+G9drNaQtCY17NTywzWd57Rzvq4sD8ABVggXmerhSm05VM-htdz88DCk4AG6aJjx0ZM5jMdDQ2E8HAEEjkcDJ-DkKgcHwcDTKY6dLiaMrpmqMGiaPH4gB03fpkPQfmXi9UsAU9Euj80On0hkYxmAgUhwGAKBCFgABaBJuCIMCvB8Lxb3vChfxvcDSC8BkR3QIA). The other problem is also that we shouldn't limit you from using `className` and I'm not sure if that's possible to achieve. With a required `className` the ideal result would be to require **either** `className` or `css` and also accept both at the same time.
